shopfloor/services/checkout.py

```diff
@@ -231,21 +236,19 @@ def _select_document_from_location(self, location, **kw):
         return self._select_picking(pickings, "select_document")
 
     def _select_document_from_package(self, package, **kw):
-        pickings = package.move_line_ids.filtered(
+        picking = package.move_line_ids.filtered(
             lambda ml: ml.state not in ("cancel", "done")
         ).mapped("picking_id")
-        if len(pickings) > 1:
+        if len(picking) > 1:
             # Filter only if we find several pickings to narrow the
             # selection to one of the good type. If we have one picking
             # of the wrong type, it will be caught in _select_picking
             # with the proper error message.
             # Side note: rather unlikely to have several transfers ready
             # and moving the same things
-            pickings = pickings.filtered(
+            picking = picking.filtered(
                 lambda p: p.picking_type_id in self.picking_types
             )
```

Comment:
IMO this selection of picking is broken here (even before these changes) if we use this checkout scenario on the first transfer to process (the Pick). This scenario has really been thought to be used on Pack, example:

- We have 2 Pick transfers assigned (same type), both reserving the same source package
- As these 2 Pick transfers reserve from the same source package, obviously their are not moving the whole pkg ("Move entire package" not enabled), so either there is no destination package (moving them as bulk), or a different destination package than the source one.
- The 2 Pack transfers that follow these 2 Picks get either no source package or the destination package set on the Pick.

Detecting the right picking record by scanning a package is not an issue if we are processing a Pack transfer, 3 possibilities:

- no package found (because no destination pkg was set on the Pick)
- package found (because the whole package has been moved, OR a new destination package has bee created during the Pick)

Now detecting the right package while working on the Pick isn't the same story: we have 2 Picks transfers reserving the same source package, which picking should be picked then? We should protect against this use case by taking the first I guess (I don't think we want to raise an error as we are at the beginning of the flow when processing the Pick :thinking: ):

```suggestion
            picking = fields.first(
                picking.filtered(
                    lambda p: p.picking_type_id in self.picking_types
                )
            )
```

Reply:
In the pickings scenario, we must always scan something that makes clear what is being processed.
In the checkout scenario, we decided to open the first matching. This was put in place when scanning raw products. So indeed, the suggestion of @sebalix to take the first is the right solution to be consistent.

However, the filtering on the picking type must always occur. This independently on the length of pickings. I would keep "pickings" instead of "picking" if you apply first in a later step, and removed the useless mapped.